File: C++/Solution1049.cpp
#include <iostream>
#include <vector>

#include <unordered_map>
#include <unordered_set>
#include <queue>

using namespace std;

class Solution1049 {
public:
    int lastStoneWeightII(vector<int>& stones) {
        int size = stones.size();
        int sum = 0;
        for(auto ele: stones){
            sum += ele;
        }
        vector<vector<int>> dp(size + 1, vector<int>(sum / 2 + 1));
        dp[0][0] = 1;
        for(int i = 0; i < size; i++){
            for(int j = 0; j <= sum / 2; j++){
                dp[i + 1][j] = dp[i][j];
                if(j >= stones[i]){
                    dp[i + 1][j] |= dp[i][j - stones[i]];
                }
            }
        }
        for(int i = sum / 2; i >= 0; i--){
            if(dp[size][i]){
                return sum - 2 * i;
            }
        }
        return 0;
    }
};

File: Java/dynamic/Solution1049.java
class Solution1049 {
    public int lastStoneWeightII(int[] stones) {
    	int sum = 0;
    	int number = stones.length;
    	for(int ele: stones){
    		sum += ele;
    	}
    	int[] dp = new int[sum / 2 + 1];
    	for(int i = 1; i <= number; i++){
    		for(int j = dp.length - 1; j >= stones[i - 1]; j--){
    			dp[j] = Math.max(dp[j], dp[j - stones[i - 1]] + stones[i - 1]);
    		}
    	}
    	return sum - 2 * dp[dp.length - 1];
    }

    public static void main(String[] args){
    	Solution1049 s = new Solution1049();
    	int[] stones = {2, 4, 1, 1};
    	System.out.println(s.lastStoneWeightII(stones));
    }
}